File: adapter/src/main/java/com/github/mrmitew/bodylog/adapter/profile_details/last_updated/presenter/LastUpdatedPresenter.java
package com.github.mrmitew.bodylog.adapter.profile_details.last_updated.presenter;

import com.github.mrmitew.bodylog.adapter.common.model.ResultState;
import com.github.mrmitew.bodylog.adapter.common.model.StateError;
import com.github.mrmitew.bodylog.adapter.common.model.UIIntent;
import com.github.mrmitew.bodylog.adapter.common.presenter.DetachableMviPresenter;
import com.github.mrmitew.bodylog.adapter.profile_common.intent.LoadProfileIntent;
import com.github.mrmitew.bodylog.adapter.profile_common.interactor.LoadProfileInteractor;
import com.github.mrmitew.bodylog.adapter.profile_details.last_updated.intent.GetProfileLastUpdatedIntent;
import com.github.mrmitew.bodylog.adapter.profile_details.last_updated.model.LastUpdatedTextState;
import com.github.mrmitew.bodylog.adapter.profile_details.last_updated.view.LastUpdatedView;
import com.jakewharton.rxrelay2.BehaviorRelay;

import javax.inject.Inject;

import io.reactivex.Observable;

public class LastUpdatedPresenter extends DetachableMviPresenter<LastUpdatedView, LastUpdatedTextState> {
    //
    // Interactors
    //

    /**
     * Loads a profile from the repository
     */
    private LoadProfileInteractor mLoadProfileInteractor;

    //
    // State relays
    //

    /*
     * State relays are subscribed to the business logic (model) and will cache (and perhaps* emit) the latest changes in the
     * business logic.
     *
     * * If the View is not attached, the relays will keep a cached state of a particular result, which
     * will be emitted as soon as the View attaches once again.
     */

    /**
     * Profile state relay
     */
    private final BehaviorRelay<ResultState> mProfileResultStateRelay;

    @Inject
    public LastUpdatedPresenter(final LoadProfileInteractor loadProfileInteractor,
                                final BehaviorRelay<ResultState> profileResultStateRelay) {
        super(null);
        mLoadProfileInteractor = loadProfileInteractor;
        mProfileResultStateRelay = profileResultStateRelay;
    }

    @Override
    protected void bindInternalIntents() {
        super.bindInternalIntents();
        mModelGateways.add(Observable.just(new LoadProfileIntent())
                .compose(mLoadProfileInteractor)
                .doOnNext(state -> System.out.println(String.format("[LAST_UPDATED] [PROFILE MODEL] (%s) : %s", state.hashCode(), state)))
                .subscribe(mProfileResultStateRelay));
    }

    @Override
    protected Observable<ResultState> createResultStateObservable(final Observable<UIIntent> uiIntentObservable) {
        return uiIntentObservable
                .publish(shared -> shared.ofType(GetProfileLastUpdatedIntent.class).flatMap(__ -> mProfileResultStateRelay));
    }

    @Override
    protected LastUpdatedTextState createViewState(final LastUpdatedTextState previousState, final ResultState resultState) {
        if (resultState instanceof LoadProfileInteractor.State) {
            if (resultState.isInProgress()) {
                // We are not going to indicate we are currently fetching new data,
                // nor will replace the old one with something, but we'll just clear the error field (if there was one)
                if (previousState.error() != StateError.Empty.INSTANCE) {
                    return previousState.toBuilder()
                            .error(StateError.Empty.INSTANCE)
                            .build();
                }
                // No error? Then, just emit the old state. No view state changes needed to be done here.
                return previousState;
            } else if (resultState.isSuccessful()) {
                return previousState.toBuilder()
                        .lastUpdated(LastUpdatedTextState.DATE_FORMAT.format(((LoadProfileInteractor.State) resultState).profile().timestamp()))
                        .error(StateError.Empty.INSTANCE)
                        .build();
            } else if (!(resultState.error() instanceof StateError.Empty)) {
                return previousState.toBuilder()
                        .lastUpdated(LastUpdatedTextState.EMPTY)
                        .error(resultState.error())
                        .build();
            }
        }

        throw new IllegalArgumentException("Unknown partial state: " + resultState);
    }

    @Override
    protected LastUpdatedTextState getInitialState() {
        return LastUpdatedTextState.Factory.idle();
    }

    @Override
    protected Observable<UIIntent> getViewIntents() {
        return mView.getProfileLastUpdatedIntent()
                .cast(UIIntent.class);
    }
}

File: adapter/src/main/java/com/github/mrmitew/bodylog/adapter/profile_details/last_updated/presenter/LastUpdatedPresenter.kt
package com.github.mrmitew.bodylog.adapter.profile_details.last_updated.presenter

import com.github.mrmitew.bodylog.adapter.common.model.ResultState
import com.github.mrmitew.bodylog.adapter.common.model.StateError
import com.github.mrmitew.bodylog.adapter.common.model.UIIntent
import com.github.mrmitew.bodylog.adapter.common.presenter.DetachableMviPresenter
import com.github.mrmitew.bodylog.adapter.profile_common.intent.LoadProfileIntent
import com.github.mrmitew.bodylog.adapter.profile_common.interactor.LoadProfileInteractor
import com.github.mrmitew.bodylog.adapter.profile_details.last_updated.intent.GetProfileLastUpdatedIntent
import com.github.mrmitew.bodylog.adapter.profile_details.last_updated.model.LastUpdatedTextState
import com.github.mrmitew.bodylog.adapter.profile_details.last_updated.view.LastUpdatedView
import com.jakewharton.rxrelay2.BehaviorRelay
import io.reactivex.Observable
import javax.inject.Inject

class LastUpdatedPresenter
@Inject constructor(private val loadProfileInteractor: LoadProfileInteractor,
                    private val profileResultStateRelay: BehaviorRelay<ResultState>) :
        DetachableMviPresenter<LastUpdatedView, LastUpdatedTextState>(LastUpdatedView.Empty()) {

    override fun initialState(): LastUpdatedTextState = LastUpdatedTextState.Factory.idle()

    override fun viewIntents(): Observable<UIIntent> =
            if (view != null) view!!.getProfileLastUpdatedIntent().cast(UIIntent::class.java) else Observable.empty<UIIntent>()

    override fun bindInternalIntents() {
        super.bindInternalIntents()
        modelGateways.add(Observable.just(LoadProfileIntent())
                .compose(loadProfileInteractor)
                .doOnNext { println("[LAST_UPDATED] [PROFILE MODEL] (${it.hashCode()}) : $it") }
                .subscribe(profileResultStateRelay))
    }

    override fun createResultStateObservable(uiIntentStream: Observable<UIIntent>): Observable<ResultState> =
            uiIntentStream.publish { shared -> shared.ofType(GetProfileLastUpdatedIntent::class.java).flatMap { profileResultStateRelay } }

    override fun createViewState(previousState: LastUpdatedTextState, resultState: ResultState): LastUpdatedTextState {
        when (resultState) {
            is LoadProfileInteractor.State ->
                when {
                    resultState.isInProgress -> // We are not going to indicate we are currently fetching new data,
                        // nor will replace the old one with something, but we'll just clear the error field (if there was one)
                        return if (previousState.error !== StateError.Empty.INSTANCE)
                            previousState.copy(error = StateError.Empty.INSTANCE)
                        else
                            previousState

                    // No error? Then, just emit the old state. No view state changes needed to be done here.
                    resultState.isSuccessful -> return previousState.copy(
                            lastUpdated = LastUpdatedTextState.Factory.DATE_FORMAT.format(resultState.profile.timestamp),
                            error = StateError.Empty.INSTANCE)

                    resultState.error !is StateError.Empty -> return previousState.copy(
                            lastUpdated = LastUpdatedTextState.Factory.DEFAULT_VALUE,
                            error = resultState.error)
                }
        }

        throw IllegalArgumentException("Unknown partial state: " + resultState)
    }
}